File: app/src/main/java/eu/kanade/tachiyomi/data/source/online/russian/Mangachan.java
package eu.kanade.tachiyomi.data.source.online.russian;

import android.content.Context;
import android.net.Uri;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import eu.kanade.tachiyomi.data.database.models.Chapter;
import eu.kanade.tachiyomi.data.database.models.Manga;
import eu.kanade.tachiyomi.data.source.Language;
import eu.kanade.tachiyomi.data.source.LanguageKt;
import eu.kanade.tachiyomi.data.source.base.Source;
import eu.kanade.tachiyomi.data.source.model.MangasPage;
import eu.kanade.tachiyomi.data.source.model.Page;
import eu.kanade.tachiyomi.util.Parser;

public class Mangachan extends Source {

    public static final String NAME = "Mangachan";
    public static final String BASE_URL = "http://mangachan.ru";
    public static final String POPULAR_MANGAS_URL = BASE_URL + "/mostfavorites";
    public static final String SEARCH_URL = BASE_URL + "/?do=search&subaction=search&story=%s";

    public Mangachan(Context context) {
        super(context);
    }

    @Override
    public Language getLang() {
        return LanguageKt.getRU();
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String getBaseUrl() {
        return BASE_URL;
    }

    @Override
    protected String getInitialPopularMangasUrl() {
        return POPULAR_MANGAS_URL;
    }

    @Override
    protected String getInitialSearchUrl(String query) {
        return String.format(SEARCH_URL, Uri.encode(query));
    }

    @Override
    protected List<Manga> parsePopularMangasFromHtml(Document parsedHtml) {
        List<Manga> mangaList = new ArrayList<>();

        for (Element currentHtmlBlock : parsedHtml.select("div.content_row")) {
            Manga manga = constructPopularMangaFromHtml(currentHtmlBlock);
            mangaList.add(manga);
        }

        return mangaList;
    }

    private Manga constructPopularMangaFromHtml(Element currentHtmlBlock) {
        Manga manga = new Manga();
        manga.source = getId();

        Element urlElement = currentHtmlBlock.getElementsByTag("h2").select("a").first();
        Element imgElement = currentHtmlBlock.getElementsByClass("manga_images").select("img").first();

        if (urlElement != null) {
            manga.setUrl(urlElement.attr("href"));
            manga.title = urlElement.text();
        }

        if (imgElement != null) {
            manga.thumbnail_url = BASE_URL + imgElement.attr("src");
        }

        return manga;
    }

    @Override
    protected String parseNextPopularMangasUrl(Document parsedHtml, MangasPage page) {
        String path = Parser.href(parsedHtml, "a:contains(Вперед)");
        return path != null ? POPULAR_MANGAS_URL + path : null;
    }

    @Override
    protected List<Manga> parseSearchFromHtml(Document parsedHtml) {
        return parsePopularMangasFromHtml(parsedHtml);
    }

    @Override
    protected String parseNextSearchUrl(Document parsedHtml, MangasPage page, String query) {
        return null;
    }

    @Override
    protected Manga parseHtmlToManga(String mangaUrl, String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);

        Element infoElement = parsedDocument.getElementsByClass("mangatitle").first();
        String description = parsedDocument.getElementById("description").text();

        Manga manga = Manga.create(mangaUrl);

        manga.author = infoElement.select("tr:eq(2) td:eq(1)").text();
        manga.genre = infoElement.select("tr:eq(5) td:eq(1)").text();
        manga.status = parseStatus(infoElement.select("tr:eq(4) td:eq(1)").text());

        manga.description = description.replaceAll("Прислать описание", "");

        manga.initialized = true;
        return manga;
    }

    private int parseStatus(String status) {
        if (status.contains("перевод продолжается")) {
            return Manga.ONGOING;
        } else if (status.contains("перевод завершен")) {
            return Manga.COMPLETED;
        } else return Manga.UNKNOWN;
    }

    @Override
    protected List<Chapter> parseHtmlToChapters(String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);
        List<Chapter> chapterList = new ArrayList<>();

        for (Element chapterElement : parsedDocument.select("table.table_cha tr:gt(1)")) {
            Chapter chapter = constructChapterFromHtmlBlock(chapterElement);
            chapterList.add(chapter);
        }
        return chapterList;
    }

    private Chapter constructChapterFromHtmlBlock(Element chapterElement) {
        Chapter chapter = Chapter.create();

        Element urlElement = chapterElement.select("a").first();
        String date = Parser.text(chapterElement, "div.date");

        if (urlElement != null) {
            chapter.name = urlElement.text();
            chapter.url = urlElement.attr("href");
        }

        if (date != null) {
            try {
                chapter.date_upload = new SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH).parse(date).getTime();
            } catch (ParseException e) { /* Ignore */ }
        }
        return chapter;
    }

    // Without this extra chapters are in the wrong place in the list
    @Override
    public void parseChapterNumber(Chapter chapter) {
        // For chapters with url like /online/254903-fairy-tail_v56_ch474.html
        String url = chapter.url.replace(".html", "");
        Pattern pattern = Pattern.compile("\\d+_ch[\\d.]+");
        Matcher matcher = pattern.matcher(url);

        if (matcher.find()) {
            String[] parts = matcher.group().split("_ch");
            chapter.chapter_number = Float.parseFloat(parts[0] + "." + AddZero(parts[1]));
        } else { // For chapters with url like /online/61216-3298.html
            String name = chapter.name;
            name = name.replaceAll("[\\s\\d\\w\\W]+v", "");
            String volume = name.substring(0, name.indexOf(" - "));
            String[] parts = name.replaceFirst("\\d+ - ", "").split(" ");

            chapter.chapter_number = Float.parseFloat(volume + "." + AddZero(parts[0]));
        }
    }

    private String AddZero(String num) {
        if (Float.parseFloat(num) < 1000f) {
            num = "0" + num.replace(".", "");
        }
        if (Float.parseFloat(num) < 100f) {
            num = "0" + num.replace(".", "");
        }
        if (Float.parseFloat(num) < 10f) {
            num = "0" + num.replace(".", "");
        }
        return num;
    }

    @Override
    protected List<String> parseHtmlToPageUrls(String unparsedHtml) {
        ArrayList<String> pages = new ArrayList<>();

        int beginIndex = unparsedHtml.indexOf("fullimg\":[");
        int endIndex = unparsedHtml.indexOf("]", beginIndex);

        String trimmedHtml = unparsedHtml.substring(beginIndex + 10, endIndex);
        trimmedHtml = trimmedHtml.replaceAll("\"", "");

        String[] pageUrls = trimmedHtml.split(",");
        for (int i = 0; i < pageUrls.length; i++) {
            pages.add("");
        }

        return pages;
    }

    @Override
    protected List<Page> parseFirstPage(List<? extends Page> pages, String unparsedHtml) {
        int beginIndex = unparsedHtml.indexOf("fullimg\":[");
        int endIndex = unparsedHtml.indexOf("]", beginIndex);

        String trimmedHtml = unparsedHtml.substring(beginIndex + 10, endIndex);
        trimmedHtml = trimmedHtml.replaceAll("\"", "");

        String[] pageUrls = trimmedHtml.split(",");
        for (int i = 0; i < pageUrls.length; i++) {
            pages.get(i).setImageUrl(pageUrls[i].replaceAll("im.?\\.", ""));
        }

        return (List<Page>) pages;
    }

    @Override
    protected String parseHtmlToImageUrl(String unparsedHtml) {
        return null;
    }
}

File: app/src/main/java/eu/kanade/tachiyomi/data/source/online/russian/Mangachan.kt
package eu.kanade.tachiyomi.data.source.online.russian

import eu.kanade.tachiyomi.data.database.models.Chapter
import eu.kanade.tachiyomi.data.database.models.Manga
import eu.kanade.tachiyomi.data.source.model.MangasPage
import eu.kanade.tachiyomi.data.source.model.Page
import eu.kanade.tachiyomi.data.source.online.ParsedOnlineSource
import eu.kanade.tachiyomi.util.asJsoup
import okhttp3.Response
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import java.text.SimpleDateFormat
import java.util.*

class Mangachan(override val id: Int) : ParsedOnlineSource() {

    override val name = "Mangachan"

    override val baseUrl = "http://mangachan.me"

    override val lang = "ru"

    override val supportsLatest = true

    override fun popularMangaInitialUrl() = "$baseUrl/mostfavorites"

    override fun latestUpdatesInitialUrl() = "$baseUrl/newestch"

    override fun searchMangaInitialUrl(query: String, filters: List<Filter<*>>): String {
        if (query.isNotEmpty()) {
            return "$baseUrl/?do=search&subaction=search&story=$query"
        } else {
            val filt = filters.filter { it.state != Filter.TriState.STATE_IGNORE }
            if (filt.isNotEmpty()) {
                var genres = ""
                filt.forEach { genres += (if (it.state == Filter.TriState.STATE_EXCLUDE) "-" else "") + (it as Genre).id + '+' }
                return "$baseUrl/tags/${genres.dropLast(1)}"
            } else {
                return "$baseUrl/?do=search&subaction=search&story=$query"
            }
        }
    }

    override fun popularMangaSelector() = "div.content_row"

    override fun latestUpdatesSelector() = "ul.area_rightNews li"

    override fun searchMangaSelector() = popularMangaSelector()

    override fun popularMangaFromElement(element: Element, manga: Manga) {
        element.select("h2 > a").first().let {
            manga.setUrlWithoutDomain(it.attr("href"))
            manga.title = it.text()
        }
    }

    override fun latestUpdatesFromElement(element: Element, manga: Manga) {
        element.select("a:nth-child(1)").first().let {
            manga.setUrlWithoutDomain(it.attr("href"))
            manga.title = it.text()
        }
    }

    override fun searchMangaFromElement(element: Element, manga: Manga) {
        popularMangaFromElement(element, manga)
    }

    override fun popularMangaNextPageSelector() = "a:contains(Вперед)"

    override fun latestUpdatesNextPageSelector() = popularMangaNextPageSelector()

    override fun searchMangaNextPageSelector() = "a:contains(Далее)"

    private fun searchGenresNextPageSelector() = popularMangaNextPageSelector()

    override fun searchMangaParse(response: Response, page: MangasPage, query: String, filters: List<Filter<*>>) {
        val document = response.asJsoup()
        for (element in document.select(searchMangaSelector())) {
            Manga.create(id).apply {
                searchMangaFromElement(element, this)
                page.mangas.add(this)
            }
        }
        val allIgnore = filters.all { it.state == Filter.TriState.STATE_IGNORE }
        searchMangaNextPageSelector().let { selector ->
            if (page.nextPageUrl.isNullOrEmpty() && allIgnore) {
                val onClick = document.select(selector).first()?.attr("onclick")
                val pageNum = onClick?.substring(23, onClick.indexOf("); return(false)"))
                page.nextPageUrl = searchMangaInitialUrl(query, emptyList()) + "&search_start=" + pageNum
            }
        }

        searchGenresNextPageSelector().let { selector ->
            if (page.nextPageUrl.isNullOrEmpty() && !allIgnore) {
                val url = document.select(selector).first()?.attr("href")
                page.nextPageUrl = searchMangaInitialUrl(query, filters) + url
            }
        }
    }

    override fun mangaDetailsParse(document: Document, manga: Manga) {
        val infoElement = document.select("table.mangatitle").first()
        val descElement = document.select("div#description").first()
        val imgElement = document.select("img#cover").first()

        manga.author = infoElement.select("tr:eq(2) > td:eq(1)").text()
        manga.genre = infoElement.select("tr:eq(5) > td:eq(1)").text()
        manga.status = parseStatus(infoElement.select("tr:eq(4) > td:eq(1)").text())
        manga.description = descElement.textNodes().first().text()
        manga.thumbnail_url = baseUrl + imgElement.attr("src")
    }

    private fun parseStatus(element: String): Int {
        when {
            element.contains("перевод завершен") -> return Manga.COMPLETED
            element.contains("перевод продолжается") -> return Manga.ONGOING
            else -> return Manga.UNKNOWN
        }
    }

    override fun chapterListSelector() = "table.table_cha tr:gt(1)"

    override fun chapterFromElement(element: Element, chapter: Chapter) {
        val urlElement = element.select("a").first()

        chapter.setUrlWithoutDomain(urlElement.attr("href"))
        chapter.name = urlElement.text()
        chapter.date_upload = element.select("div.date").first()?.text()?.let {
            SimpleDateFormat("yyyy-MM-dd", Locale.US).parse(it).time
        } ?: 0
    }

    override fun pageListParse(response: Response, pages: MutableList<Page>) {
        val html = response.body().string()
        val beginIndex = html.indexOf("fullimg\":[") + 10
        val endIndex = html.indexOf(",]", beginIndex)
        val trimmedHtml = html.substring(beginIndex, endIndex).replace("\"", "")
        val pageUrls = trimmedHtml.split(',')

        pageUrls.mapIndexedTo(pages) { i, url -> Page(i, "", url) }
    }

    override fun pageListParse(document: Document, pages: MutableList<Page>) {
    }

    override fun imageUrlParse(document: Document) = ""

    private class Genre(name: String, val id: String = name.replace(' ', '_')) : Filter.TriState(name)

    /* [...document.querySelectorAll("li.sidetag > a:nth-child(1)")].map((el,i) =>
    *  { const link=el.getAttribute('href');const id=link.substr(6,link.length);
    *  return `Genre("${id.replace("_", " ")}")` }).join(',\n')
    *  on http://mangachan.me/
    */
    override fun getFilterList(): List<Filter<*>> = listOf(
            Genre("18 плюс"),
            Genre("bdsm"),
            Genre("арт"),
            Genre("биография"),
            Genre("боевик"),
            Genre("боевые искусства"),
            Genre("вампиры"),
            Genre("веб"),
            Genre("гарем"),
            Genre("гендерная интрига"),
            Genre("героическое фэнтези"),
            Genre("детектив"),
            Genre("дзёсэй"),
            Genre("додзинси"),
            Genre("драма"),
            Genre("игра"),
            Genre("инцест"),
            Genre("искусство"),
            Genre("история"),
            Genre("киберпанк"),
            Genre("кодомо"),
            Genre("комедия"),
            Genre("литРПГ"),
            Genre("магия"),
            Genre("махо-сёдзё"),
            Genre("меха"),
            Genre("мистика"),
            Genre("музыка"),
            Genre("научная фантастика"),
            Genre("повседневность"),
            Genre("постапокалиптика"),
            Genre("приключения"),
            Genre("психология"),
            Genre("романтика"),
            Genre("самурайский боевик"),
            Genre("сборник"),
            Genre("сверхъестественное"),
            Genre("сказка"),
            Genre("спорт"),
            Genre("супергерои"),
            Genre("сэйнэн"),
            Genre("сёдзё"),
            Genre("сёдзё-ай"),
            Genre("сёнэн"),
            Genre("сёнэн-ай"),
            Genre("тентакли"),
            Genre("трагедия"),
            Genre("триллер"),
            Genre("ужасы"),
            Genre("фантастика"),
            Genre("фурри"),
            Genre("фэнтези"),
            Genre("школа"),
            Genre("эротика"),
            Genre("юри"),
            Genre("яой"),
            Genre("ёнкома")
    )
}